### ISSUE TITLE
OrderBy Post Properties Value

### ISSUE THREAD

**Issue by CentricStraaten** [PiranhaCMS/Piranha#509]
I have added extra properties to a Post Type. Now i like to select Posts based on these property values. 
I tried following code but received an error?

`            using (var db = new Piranha.DataContext())
            {
                calendarItems =db.Posts
                    .Include(p => p.CreatedBy)
                    .Include(p => p.Properties)
                    .Where(p => p.Template.Name == "Agenda")
                    .OrderByDescending(p => p.Properties.ByName("Datum").Value)
                    .ToList();
            }
`
How do i solve this?

**Comment by tidyui** [comment]
The following statement:

    ByName("Datum")

is a code extension and cannot be translated to SQL. The only way around it is to call the method after executing the query by calling it after `ToList()`.

    using (var db = new Piranha.DataContext()) { 
      calendarItems = db.Posts
        .Include(p => p.CreatedBy)
        .Include(p => p.Properties)
        .Where(p => p.Template.Name == "Agenda")
        .ToList() // This forces the sql query to be executed
        .OrderByDescending(p => p.Properties.ByName("Datum").Value)
        .ToList(); 
    }

Regards

**Comment by CentricStraaten** [comment]
If I follow your instruction I would  order the records after fetching all records from the database? That is not very efficient.

**Comment by tidyui** [comment]
No you wouldn't... The `ToList()` call is made **after** the `Where()` statement which means you're only fetching the records that has the template name "Agenda". You only perform the sorting in memory.

**Comment by tidyui** [comment]
When reading through the code again I saw you were also using `ByName("Datum")` which also can't be transformed to SQL. Since property values are just strings I tried to rewrite your query to plain Entity Framework code like this:

    using (var db = new Piranha.DataContext()) { 
      calendarItems = db.Posts
        .Include(p => p.CreatedBy)
        .Include(p => p.Properties)
        .Where(p => p.Template.Name == "Agenda")
        .OrderByDescending(p => p.Properties.FirstOrDefault(pr => pr.Name == "Datum").Value)
        .ToList(); 
    }

Unfortunately this throws the following exception:

`The text, ntext, and image data types cannot be compared or sorted, except when using IS NULL or LIKE operator.`

This is something built into Entity Framework and nothing I can't work around in the application layer. So I would still suggesting getting the posts back from the database and then sorting them in memory. You would still have to have a pretty large amount of posts for it to impact performance. And if you're going to have a lot of post I still suggest you cache the result in the web application for X minutes so the data fetching & sorting is not executed on every request.

Regards
